### PR TITLE
Mp imagenet scheduler

### PR DESCRIPTION
**Pull request by zcain117**
Add learning rate scheduler to multiprocessing version of resnet50 on imagenet

### CODE REVIEW

File: test/test_train_mp_imagenet.py [hunk 1]
```diff
@@ -132,13 +151,30 @@ def train_imagenet():
 
   torch.manual_seed(42)
 
+  devices = (
```
Comment (dlibenzi) [inline test/test_train_mp_imagenet.py]: This will always return 1 device.
IOW, do not use that API for this.

File: test/test_train_mp_imagenet.py [hunk 2]
```diff
@@ -132,13 +151,30 @@ def train_imagenet():
...
+          max_devices=FLAGS.num_cores) if FLAGS.num_cores != 0 else [])
   device = xm.xla_device()
   model = get_model_property('model_fn')().to(device)
+  writer = SummaryWriter(log_dir=FLAGS.logdir) if FLAGS.logdir else None
   optimizer = optim.SGD(
       model.parameters(),
       lr=FLAGS.lr,
       momentum=FLAGS.momentum,
       weight_decay=5e-4)
+  total_num_devices = len(
```
Comment (dlibenzi) [inline test/test_train_mp_imagenet.py]: You need to use `xm.get_world_size()` to get the total number of devices.
Even the non MP one, will not work correctly on POD.

File: test/test_train_mp_imagenet.py [hunk 3]
```diff
@@ -132,13 +151,30 @@ def train_imagenet():
...
+  num_training_steps_per_epoch = train_dataset_len // (
+      FLAGS.batch_size * total_num_devices)
+  lr_scheduler = schedulers.wrap_optimizer_with_scheduler(
+      optimizer,
+      scheduler_type=getattr(FLAGS, 'lr_scheduler_type', None),
+      scheduler_divisor=getattr(FLAGS, 'lr_scheduler_divisor', None),
+      scheduler_divide_every_n_epochs=getattr(
+          FLAGS, 'lr_scheduler_divide_every_n_epochs', None),
+      num_steps_per_epoch=num_training_steps_per_epoch,
+      summary_writer=writer if test_utils.is_first_device(
```
Comment (dlibenzi) [inline test/test_train_mp_imagenet.py]: Here there is only one device.
So maybe:
```Python
test_utils.is_first_device(device, [device])

Comment (zcain117) [reply test/test_train_mp_imagenet.py]: done

File: test/test_train_mp_imagenet.py (hunk 2 above)
Comment (zcain117) [reply test/test_train_mp_imagenet.py]: let me know if it looks right now

File: test/test_train_mp_imagenet.py (hunk 1 above)
Comment (zcain117) [reply test/test_train_mp_imagenet.py]: removed this line

File: test/test_train_mp_imagenet.py [hunk 4]
```diff
@@ -134,23 +153,38 @@ def train_imagenet():
...
+          device, [], multiprocess=True) else None)
   loss_fn = nn.CrossEntropyLoss()
 
   def train_loop_fn(loader):
     tracker = xm.RateTracker()
     model.train()
     for x, (data, target) in loader:
+      print("LR:{}".format(optimizer.param_groups[0]['lr']))
```
Comment (zcain117) [inline test/test_train_mp_imagenet.py]: I'll remove this before submitting

File: test/test_train_mp_imagenet.py (hunk 3 above)
Comment (zcain117) [reply test/test_train_mp_imagenet.py]: On second thought: changed is_first_device to have a multiprocess flag to make it more obvious what is going on

File: test/test_utils.py [hunk 5]
```diff
@@ -53,8 +53,10 @@ def is_first_device(current_device, devices):
     current_device: instance of `torch.device`.
     devices: list of device names (strings), e.g. output of
         torch_xla_py.xla_model.get_xla_supported_devices().
+    multiprocess: If true, device is ignored and only ordinal is considered.
```
Comment (dlibenzi) [inline test/test_utils.py]: Why?

Comment (zcain117) [reply test/test_utils.py]: I wanted to be explicit. I thought this would be more clear than `is_first_device(device, [str(device)])`

Comment (zcain117) [reply test/test_utils.py]: The `device` arg seems irrelevant on multiprocess. It is always "xla:0"

Comment (dlibenzi) [reply test/test_utils.py]: The semantics are the same.
You pass in the local device, and the list of all local devices.

File: test/test_train_mp_imagenet.py (hunk 3 above)
Comment (zcain117) [reply test/test_train_mp_imagenet.py]: On third thought: changed back to the original :)
I agree with Davide's point

File: test/test_utils.py (hunk 5 above)
Comment (zcain117) [reply test/test_utils.py]: That's true. Switched back to original version

File: test/test_train_mp_imagenet.py (hunk 4 above)
Comment (zcain117) [reply test/test_train_mp_imagenet.py]: removed

**Comment by zcain117** [comment]
Running locally and LR seems to update correctly. I'll let run a few epochs and let circleci finish before merge
